anton_gadzikovskiy/graph_generator.cpp

```diff
@@ -30,38 +38,115 @@ class Graph {
   };
 
   struct Edge {
+    enum class Color { Gray, Green, Yellow, Red };
+
     const VertexId from_vertex_id;
     const VertexId to_vertex_id;
     const EdgeId id;
+    const Color color;
 
     Edge(const EdgeId& _id,
          const VertexId& _from_vertex_id,
-         const VertexId& _to_vertex_id)
+         const VertexId& _to_vertex_id,
+         const Color& _color)
         : id(_id),
           from_vertex_id(_from_vertex_id),
-          to_vertex_id(_to_vertex_id) {}
+          to_vertex_id(_to_vertex_id),
+          color(_color) {}
   };
 
   void add_vertex() {
+    if (vertices_on_depth_.empty()) {
+      vertices_on_depth_.emplace(0, vector<VertexId>());
+    }
     const VertexId new_vertex_id = get_new_vertex_id();
     vertices_.emplace(new_vertex_id, new_vertex_id);
+    vertices_on_depth_.at(0).push_back(new_vertex_id);
   }
 
   void add_edge(const VertexId& from_vertex_id, const VertexId& to_vertex_id) {
     const EdgeId new_edge_id = get_new_edge_id();
-    edges_.emplace_back(new_edge_id, from_vertex_id, to_vertex_id);
+    const auto& new_edge_color =
+        define_edge_color(from_vertex_id, to_vertex_id);
+    edges_.emplace_back(new_edge_id, from_vertex_id, to_vertex_id,
+                        new_edge_color);
     vertices_.at(from_vertex_id).add_edge_id(new_edge_id);
-    vertices_.at(to_vertex_id).add_edge_id(new_edge_id);
+    if (new_edge_color != Edge::Color::Green) {
+      vertices_.at(to_vertex_id).add_edge_id(new_edge_id);
+    }
+    if (new_edge_color == Edge::Color::Gray) {
+      const auto new_vertex_depth = vertices_.at(from_vertex_id).depth + 1;
+      vertices_.at(to_vertex_id).depth = new_vertex_depth;
+      if (vertices_on_depth_.size() - 1 < new_vertex_depth) {
+        vertices_on_depth_.emplace(new_vertex_depth, vector<VertexId>());
```

Comment:
А сам `id` на новую глубину положить?
```suggestion
        vertices_on_depth_.emplace(new_vertex_depth, vector<VertexId>({to_vertex_id}));
```